### PR TITLE
Safe Tracker

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Include relevant motivation and context. A link to the issue is enough if it contains all the relevant 
information.

## Changes proposed in this pull request

Include a summary of the change.

## Test plan

Please describe the tests cases that you ran to verify your changes. Add further instructions on 
how to run them if needed (i.e. migration / deployment scripts, env vars, etc).

## Checklist

- [ ] Rebased the base branch
- [ ] Attached corresponding Github issue
- [ ] Prefixed the name with the type of change (i.e. feat, chore, test)
- [ ] Performed a self-review of my own code
- [ ] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [ ] Didn't generate new warnings
- [ ] Didn't generate failures on existing tests
- [ ] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/capital/SafeTracker.sol [hunk 1]
```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "../../interfaces/ISafeTracker.sol";
+import "../../interfaces/IPool.sol";
+import "../../interfaces/ISwapOperator.sol";
+import "../../interfaces/IPriceFeedOracle.sol";
+import "../../abstract/MasterAwareV2.sol";
+import "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:10]: oz import goes at the top

Comment (shark0der) [reply contracts/modules/capital/SafeTracker.sol:10]: also, except for the oz import, sort the imports alphabetically

File: contracts/modules/capital/Pool.sol [hunk 2]
```diff
@@ -15,6 +15,7 @@ import "../../interfaces/IPriceFeedOracle.sol";
 import "../../interfaces/ISwapOperator.sol";
 import "../../libraries/Math.sol";
 import "../../libraries/SafeUintCast.sol";
+import {ISafeTracker} from "../../interfaces/ISafeTracker.sol";
```
Comment (shark0der) [inline contracts/modules/capital/Pool.sol:18]: this code will be deleted anyway but there's no need for `{}` here

File: contracts/modules/capital/SafeTracker.sol [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+
+contract SafeTracker is ISafeTracker, MasterAwareV2 {
+
+  address public _swapOperator;
+  uint public investedUSDC;
+  uint public transferAmount;
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:17]: requested* amount

File: contracts/modules/capital/SafeTracker.sol [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    return _calculateBalance();
+  }
+
+  /**
+  * @dev Gets the balance of the safe
+  * @return An uint256 representing the amount of the safe.
+  */
+  function balanceOf(address) external view onlyPool returns (uint256) {
+    return _calculateBalance();
+  }
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:71]: 1. Anyone should be able to call `balanceOf`, no need for the modifier.
2. It should report that balance only for the pool contract address

File: contracts/modules/capital/SafeTracker.sol [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+  * @dev Updates invested USDC
+  */
+  function updateInvestedUSDC(uint _investedUSDC) external {
+    require(msg.sender == safe, "SafeTracker: not safe");
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:77]: Let's have custom errors please

File: contracts/modules/capital/SafeTracker.sol [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    investedUSDC = _investedUSDC;
+  }
+
+  /**
+  * @dev emits Transfer event only if it's called by Pool or SwapOperator
+  */
+  function transfer(address, uint256) external onlyPoolAndSwapOperator returns (bool){
+    emit Transfer(address(0), address(0), 0); // add arguments
+    return true;
+  }
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:87]: Calling transfer with a zero amount should not revert. It is not the case to use a modifier here. Do the checks explicitly inside the function.

File: contracts/modules/capital/SafeTracker.sol [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+  * @dev emits Transfer event only if it's called by Pool or SwapOperator
+  */
+  function transferFrom(address, address, uint256) external override onlyPoolAndSwapOperator returns(bool) {
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:92]: same here

File: contracts/modules/capital/SafeTracker.sol [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    emit Transfer(address(0), address(0), 0); // add arguments
+    return true;
+  }
+
+  function allowance(address, address) external pure returns (uint256){
+    return 0;
+  }
+
+  function approve(address, uint256) external view override onlyPoolAndSwapOperator returns (bool){
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:101]: remove the modifier

Comment (shark0der) [reply contracts/modules/capital/SafeTracker.sol:101]: would make sense to revert here

File: contracts/modules/capital/SafeTracker.sol [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    // usdc debt (borrowed usdc)
+    uint usdcDebtAmount = debtUsdc.balanceOf(safe);
+    uint usdcDebtInEth = priceFeedOracle.getEthForAsset(address(usdc), usdcDebtAmount);
+
+    return ethAmount + usdcInEth - usdcDebtInEth;
+  }
+
+  function requestTransfer(uint amount) external onlySafe {
+    transferAmount = amount;
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:128]: `s/transferAmount/requestedAmount`

Comment (shark0der) [reply contracts/modules/capital/SafeTracker.sol:128]: (code will be moved to SO)

File: contracts/modules/capital/SafeTracker.sol [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  function executeTransfer() external onlySwapOperator {
+    transferAmount = 0;
+  }
+
+  function getSafeAddress() external view returns (address) {
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:135]: Seems a bit redundant as safe is already public. any specific reason why you want this here?

File: contracts/modules/capital/SafeTracker.sol [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  function swapOperator() internal view returns (ISwapOperator) {
+    return ISwapOperator(internalContracts[uint(ID.P1)]);
+  }
+
+  function updateSwapOperatorAddress(address newAddress) external onlyGovernance {
+    if (_swapOperator != address(0)) {
+      require(swapOperator().orderInProgress(), 'SafeTracker: Cancel all swaps before changing swapOperator');
+    }
+    _swapOperator = newAddress;
+  }
```
Comment (shark0der) [inline contracts/modules/capital/SafeTracker.sol:154]: I'm under the impression you will not need the swap operator address anymore based on the comments above, but if it's still needed SO should be read from the pool contract - that's the source of truth

File: contracts/modules/capital/SwapOperator.sol [hunk 12]
```diff
@@ -504,9 +513,22 @@ contract SwapOperator {
...
+    address safe = safeTracker.safe();
+    _pool().transferAssetToSwapOperator(ETH, amount);
+    transferAssetTo(ETH, safe, amount);
+    safeTracker.executeTransfer();
+  }
+
   function recoverAsset(address assetAddress, address receiver) public onlyController {
 
-    // Validate there's no current cow swap order going on
+    // Validate there's no current cow wap order going on
```
Comment (shark0der) [inline contracts/modules/capital/SwapOperator.sol:544]: no, lol

File: contracts/modules/capital/SwapOperator.sol [hunk 13]
```diff
@@ -504,9 +519,29 @@ contract SwapOperator {
...
+  }
+
+  /**
+   * @dev Transfer request amount of ETH to the safe
+   */
+  function transferRequestedETH() external onlyController returns (bool){
+    uint amount = safeRequestedAmount;
+    _pool().transferAssetToSwapOperator(ETH, amount);
+    transferAssetTo(ETH, safe, amount);
+    safeRequestedAmount = 0;
```
Comment (shark0der) [inline contracts/modules/capital/SwapOperator.sol:537]: zero it out right after reading:
```
    uint amount = safeRequestedAmount;
    safeRequestedAmount = 0;
    _pool().transferAssetToSwapOperator(ETH, amount);
    transferAssetTo(ETH, safe, amount);
    ```

File: contracts/modules/capital/SwapOperator.sol [hunk 14]
```diff
@@ -504,9 +519,29 @@ contract SwapOperator {
...
+   * @dev Request amount of ETH to be transferred to the safe
+   */
+  function requestETH(uint amount) external onlySafe {
+    safeRequestedAmount = amount;
+  }
+
+  /**
+   * @dev Transfer request amount of ETH to the safe
+   */
+  function transferRequestedETH() external onlyController returns (bool){
```
Comment (shark0der) [inline contracts/modules/capital/SwapOperator.sol:533]: remove the return param, it has absolutely no purpose here

File: contracts/modules/capital/SwapOperator.sol (hunk 12 above)
Comment (MilGard91) [reply contracts/modules/capital/SwapOperator.sol:544]: accident

File: contracts/modules/capital/SafeTracker.sol (hunk 10 above)
Comment (MilGard91) [reply contracts/modules/capital/SafeTracker.sol]: Cleaned it up already

File: contracts/modules/capital/SafeTracker.sol (hunk 8 above)
Comment (MilGard91) [reply contracts/modules/capital/SafeTracker.sol:96]: We said we would revert but first we will check if the sender is pool or swapOperator...

File: contracts/interfaces/IPool.sol [hunk 15]
```diff
@@ -20,6 +20,8 @@ struct Asset {
 
 interface IPool {
 
+  function swapOperator() external view returns (address);
```
Comment (roxdanila) [inline contracts/interfaces/IPool.sol:23]: leftover?

File: contracts/modules/capital/SafeTracker.sol [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    if (account != address(pool())) {
+      return 0;
+    }
+    return _calculateBalance();
+  }
+
+  /**
+  * @dev Updates invested USDC
+  */
+  function updateInvestedUSDC(uint _investedUSDC) external {
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:55]: maybe rename to `updateCoverReInvestmentUSDC(amount)`?

File: contracts/modules/capital/SafeTracker.sol [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+import "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import "../../abstract/MasterAwareV2.sol";
+import "../../interfaces/IPool.sol";
+import "../../interfaces/IPriceFeedOracle.sol";
+import "../../interfaces/ISafeTracker.sol";
+
+contract SafeTracker is ISafeTracker, MasterAwareV2 {
+
+  uint public investedUSDC;
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:13]: rename to `coverReInvestmentUSDC` or similar

File: contracts/modules/capital/SafeTracker.sol [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+  function updateInvestedUSDC(uint _investedUSDC) external {
+    require(msg.sender == safe, "SafeTracker: not safe");
+    investedUSDC = _investedUSDC;
+  }
+
+  /**
+  * @dev emits Transfer event only if it's called by Pool or SwapOperator
+  */
+  function transfer(address, uint256 amount) external returns (bool){
+    if (amount == 0 || msg.sender == internalContracts[uint(ID.P1)] || msg.sender == pool().swapOperator()) {
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:64]: you mean amount !== 0?

File: contracts/modules/capital/SafeTracker.sol [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+      emit Transfer(address(0), address(0), 0); // add arguments
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:65]: todo

File: contracts/modules/capital/SafeTracker.sol [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+  */
+  function updateInvestedUSDC(uint _investedUSDC) external {
+    require(msg.sender == safe, "SafeTracker: not safe");
+    investedUSDC = _investedUSDC;
+  }
+
+  /**
+  * @dev emits Transfer event only if it's called by Pool or SwapOperator
+  */
+  function transfer(address, uint256 amount) external returns (bool){
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:63]: nit: missing space - i.e. `(bool) {`

File: contracts/modules/capital/SafeTracker.sol [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+  /**
+  * @dev Fetches all necessary information about the tokens that are used in the safe and calculates the balance
+  * @return balance ETH value of the safe.
+  */
+  function _calculateBalance() internal view returns (uint256 balance) {
+
+    // eth in the safe and aweth balance, aweth is 1:1 to eth
+    uint ethAmount = address(safe).balance + aweth.balanceOf(safe);
+
+    IPriceFeedOracle priceFeedOracle = pool().priceFeedOracle();
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:99]: you need to add to PriceFeedOracle the oracle for usdc

File: contracts/modules/capital/SafeTracker.sol [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+
+    // usdc actually in the safe
+    uint usdcAmount = usdc.balanceOf(safe) + investedUSDC;
+    uint usdcInEth = priceFeedOracle.getEthForAsset(address(usdc), usdcAmount);
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:103]: rename to `usdcValueInEth`

File: contracts/modules/capital/SafeTracker.sol [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+
+    // usdc debt (borrowed usdc)
+    uint usdcDebtAmount = debtUsdc.balanceOf(safe);
+    uint usdcDebtInEth = priceFeedOracle.getEthForAsset(address(usdc), usdcDebtAmount);
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:107]: `debtUsdcValueInEth``

File: contracts/modules/capital/SafeTracker.sol [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+    uint ethAmount = address(safe).balance + aweth.balanceOf(safe);
+
+    IPriceFeedOracle priceFeedOracle = pool().priceFeedOracle();
+
+    // usdc actually in the safe
+    uint usdcAmount = usdc.balanceOf(safe) + investedUSDC;
+    uint usdcInEth = priceFeedOracle.getEthForAsset(address(usdc), usdcAmount);
+
+    // usdc debt (borrowed usdc)
+    uint usdcDebtAmount = debtUsdc.balanceOf(safe);
```
Comment (roxdanila) [inline contracts/modules/capital/SafeTracker.sol:106]: `debtUsdcAmount` - or similar, to be consistent with `IERC20 public immutable debtUsdc;` above

File: contracts/interfaces/IPool.sol (hunk 15 above)
Comment (MilGard91) [reply contracts/interfaces/IPool.sol:23]: I'm using this in `SafeTracker` to check who is calling the function

File: contracts/modules/capital/SafeTracker.sol (hunk 18 above)
Comment (MilGard91) [reply contracts/modules/capital/SafeTracker.sol:64]: If the amount is 0, it should not revert. You can see Anatol's comment about this.

File: test/integration/setup.js [hunk 25]
```diff
@@ -121,33 +147,40 @@ async function setup() {
...
+    priceFeedOracleAssets.map(a => a.decimals),
   ]);
 
-  // placeholder is swapped with the actual one after master is initialized
-  const swapOperatorPlaceholder = { address: AddressZero };
-
-  const legacyPool = await ethers.deployContract(
-    'LegacyPool',
-    [master, priceFeedOracle, swapOperatorPlaceholder, dai, stETH, enzymeVault, tk].map(c => c.address),
-  );
```
Comment (rackstar) [inline test/integration/setup.js:136]: does anyone know why we have the `swapOperatorPlaceholder`? all tests are passing by passing in the `swapOperator` directly.

I also found out that the pool swapOperator was still the zero address. so i elected to remove it all together to simplify the code

File: test/integration/setup.js [hunk 26]
```diff
@@ -250,14 +282,16 @@ async function setup() {
...
 
-  await legacyPool.updateAddressParameters(toBytes8('SWP_OP'), swapOperator.address);
+  // Manually add pool assets that are not automatically added via LegacyPool constructor
   await legacyPool.addAsset(
     usdc.address,
     true,
     parseUnits('1000000', usdcDecimals),
     parseUnits('2000000', usdcDecimals),
     250,
   );
```
Comment (rackstar) [inline test/integration/setup.js:292]: i checked the [Pool.getAssets](https://etherscan.io/readContract?m=light&a=0xcafeaBED7e0653aFe9674A3ad862b78DB3F36e60&n=mainnet&v=0xcafeaBED7e0653aFe9674A3ad862b78DB3F36e60#readCollapse7) and it does not contain USDC as an asset.

I tried removing this line but it breaks `cover.setProducts`.

Should we bother aligning this with the deployed contract or is there no need?

File: test/integration/setup.js [hunk 27]
```diff
@@ -43,6 +44,9 @@ async function setup() {
   const INITIAL_SPOT_PRICE_B = parseEther('0.0152');
 
   // deploy external contracts
+  const gnosisSafe = await ethers.deployContract('ERC20Mock');
```
[the remark this answers is not in the document]
Comment (rackstar) [reply test/integration/setup.js:47]: Miljan suggested to us Multicall contract

File: contracts/interfaces/IPool.sol (hunk 15 above)
Comment (shark0der) [reply contracts/interfaces/IPool.sol:23]: Would this require a redeploy? Or is that view fn present in the pool but only missing in the interface?

File: contracts/mocks/SafeTracker/SafeTrackerMockPool.sol [hunk 28]
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "../common/PoolMock.sol";
+import "../../interfaces/IPriceFeedOracle.sol";
+
+contract SafeTrackerMockPool is PoolMock {
```
Comment (shark0der) [inline contracts/mocks/SafeTracker/SafeTrackerMockPool.sol:8]: Rename to `STMockPool.sol` please. It's a convention we're trying to adapt to but it's not entirely adopted

File: contracts/mocks/SafeTracker/SafeTrackerSwapOperatorMock.sol [hunk 29]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "../../interfaces/ISwapOperator.sol";
+
+contract SafeTrackerSwapOperatorMock is ISwapOperator {
```
Comment (shark0der) [inline contracts/mocks/SafeTracker/SafeTrackerSwapOperatorMock.sol:7]: STSwapOperatorMock

File: contracts/mocks/common/PriceFeedOracleMock.sol [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+contract PriceFeedOracleMock is IPriceFeedOracle {
+
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  mapping(address => OracleAsset) public assets;
+
+  uint public ethRate;
+
+  constructor(uint _ethRate) {
+    ethRate = _ethRate;
+  }
```
Comment (shark0der) [inline contracts/mocks/common/PriceFeedOracleMock.sol:16]: At this point we can have a mapping asset => assetRate. Only add if helps

**Pull request closed without merging**